### ISSUE TITLE
configure fails to find libasr

### ISSUE THREAD

**Issue by edgarpettijohn** [OpenSMTPD/OpenSMTPD#1014]
TLDR
Configure can't seem to find libasr.

localhost$ uname -a
NetBSD localhost 8.1 NetBSD 8.1 (GENERIC) #0: Fri May 31 08:43:59 UTC 2019  mkrepro@mkrepro.NetBSD.org:/usr/src/sys/arch/amd64/compile/GENERIC amd64

Installed latest libasr and libressl from source.

localhost$ ls -l /usr/local/lib
total 58712
-rw-r--r--  1 root  wheel    361952 Nov 30 23:17 libasr.a
-rwxr-xr-x  1 root  wheel       907 Nov 30 23:17 libasr.la
lrwxr-xr-x  1 root  wheel        13 Nov 30 23:17 libasr.so -> libasr.so.0.3
lrwxr-xr-x  1 root  wheel        13 Nov 30 23:17 libasr.so.0 -> libasr.so.0.3
-rwxr-xr-x  1 root  wheel    175624 Nov 30 23:17 libasr.so.0.3
-rw-r--r--  1 root  wheel  14954334 Nov 30 23:15 libcrypto.a
-rw-r--r--  1 root  wheel       943 Nov 30 23:15 libcrypto.la
lrwxr-xr-x  1 root  wheel        17 Nov 30 23:15 libcrypto.so -> libcrypto.so.45.5
lrwxr-xr-x  1 root  wheel        17 Nov 30 23:15 libcrypto.so.45 -> libcrypto.so.45.5
-rw-r--r--  1 root  wheel   7974944 Nov 30 23:15 libcrypto.so.45.5
-rw-r--r--  1 root  wheel   3615076 Nov 30 23:15 libssl.a
-rw-r--r--  1 root  wheel       950 Nov 30 23:15 libssl.la
lrwxr-xr-x  1 root  wheel        14 Nov 30 23:15 libssl.so -> libssl.so.47.6
lrwxr-xr-x  1 root  wheel        14 Nov 30 23:15 libssl.so.47 -> libssl.so.47.6
-rw-r--r--  1 root  wheel   1703368 Nov 30 23:15 libssl.so.47.6
-rw-r--r--  1 root  wheel    687112 Nov 30 23:15 libtls.a
-rw-r--r--  1 root  wheel       975 Nov 30 23:15 libtls.la
lrwxr-xr-x  1 root  wheel        14 Nov 30 23:15 libtls.so -> libtls.so.19.7
lrwxr-xr-x  1 root  wheel        14 Nov 30 23:15 libtls.so.19 -> libtls.so.19.7
-rw-r--r--  1 root  wheel    335216 Nov 30 23:15 libtls.so.19.7
drwxr-xr-x  2 root  wheel       512 Nov 30 23:15 pkgconfig

localhost$ ./configure --with-libasr=/usr/local --with-libssl=/usr/local
lots of output....
checking for asr.h... yes
checking for asr_run in -lasr... no
configure: error: *** libasr missing - please install first or check config.log ***

localhost$ grep asr config.log
  $ ./configure --with-libasr=/usr/local --with-libssl=/usr/local
configure:16544: checking for library containing event_asr_run
/home/edgar/opensmtpd/conftest.c:118: undefined reference to `event_asr_run'
| char event_asr_run ();
| return event_asr_run ();
/home/edgar/opensmtpd/conftest.c:118: undefined reference to `event_asr_run'
| char event_asr_run ();
| return event_asr_run ();
configure:19150: checking for asr.h
configure:19161: checking for asr_run in -lasr
configure:19186: gcc -o conftest -g -O2  -fPIC -DPIC -Wall -Wpointer-arith -Wuninitialized -Wsign-compare -Wformat-security -Wsizeof-pointer-memaccess -Wno-pointer-sign -Wno-unused-result -fno-strict-aliasing -fno-builtin-memset  -I/usr/local/include  -L/usr/local/lib -R/usr/local/lib  conftest.c -lasr  -lcrypt  >&5
/usr/local/lib/libasr.so: undefined reference to `recallocarray'
| char asr_run ();
| return asr_run ();
/home/edgar/opensmtpd/conftest.c:191: undefined reference to `asr_run'
| char asr_run ();
| return asr_run ();
configure:19239: error: *** libasr missing - please install first or check config.log ***
ac_cv_header_asr_h=yes
ac_cv_lib_asr_asr_run=no
ac_cv_search_event_asr_run=no
ASR_LIB='-lasr'

Ignoring the failure:
localhost$ make
make  all-recursive
Making all in openbsd-compat
gcc -DHAVE_CONFIG_H -I. -I..  -I../smtpd -I../openbsd-compat  -I../openbsd-compat/err_h  -I/usr/local/include -I/usr/include -I/usr/pkg/include  -g -O2  -fPIC -DPIC -Wall -Wpointer-arith -Wuninitialized -Wsign-compare -Wformat-security -Wsizeof-pointer-memaccess -Wno-pointer-sign -Wno-unused-result -fno-strict-aliasing -fno-builtin-memset -c -o event_asr_run.o event_asr_run.c
In file included from event_asr_run.c:26:0:
/usr/include/stdlib.h:270:27: error: macro "arc4random_stir" passed 1 arguments, but takes just 0
 void  arc4random_stir(void);
                           ^
*** Error code 1

Stop.
make[2]: stopped in /home/edgar/opensmtpd/openbsd-compat
*** Error code 1

Stop.
make[1]: stopped in /home/edgar/opensmtpd
*** Error code 1

Stop.
make: stopped in /home/edgar/opensmtpd

**Comment by edgarpettijohn** [comment]
Using libasr from pkgsrc it now completes configure, but make fails.

localhost$ ./configure --with-libasr=/usr/pkg --with-libssl=/usr/local
localhost$ make
make  all-recursive
Making all in openbsd-compat
gcc -DHAVE_CONFIG_H -I. -I..  -I../smtpd -I../openbsd-compat  -I../openbsd-compat/err_h  -I/usr/local/include -I/usr/include -I/usr/pkg/include  -g -O2  -fPIC -DPIC -Wall -Wpointer-arith -Wuninitialized -Wsign-compare -Wformat-security -Wsizeof-pointer-memaccess -Wno-pointer-sign -Wno-unused-result -fno-strict-aliasing -fno-builtin-memset -c -o event_asr_run.o event_asr_run.c
In file included from event_asr_run.c:26:0:
/usr/include/stdlib.h:270:27: error: macro "arc4random_stir" passed 1 arguments, but takes just 0
 void  arc4random_stir(void);
                           ^
*** Error code 1

Stop.
make[2]: stopped in /home/edgar/opensmtpd/openbsd-compat
*** Error code 1

Stop.
make[1]: stopped in /home/edgar/opensmtpd
*** Error code 1

Stop.
make: stopped in /home/edgar/opensmtpd

**Comment by poolpOrg** [comment]
@edgarpettijohn what version are you trying to build ?

**Comment by edgarpettijohn** [comment]
github checkout from yesterday

**Comment by poolpOrg** [comment]
Lets see if @ngortheone manages to setup a NetBSD CI now :-p

**Comment by ngortheone** [comment]
@poolpOrg @edgarpettijohn 
I'd love to setup NetBSD build, but I don't know where :) SourceHut does not provide NetBSD workers (yet) https://man.sr.ht/builds.sr.ht/compatibility.md

If you know of any CI platforms that have NetBSD - then I will gladly try!

**Comment by edgarpettijohn** [comment]
The native libevent is libevent2 linked with openssl. So I rebuilt libevent and installed to /usr/local and tried building opensmtpd again against the new libevent with the same result. I wish I knew where arc4random_stir was being called with an argument. That may help figure things out.

**Comment by edgarpettijohn** [comment]
This definantly isn't the correct solution, but adding the following to stdlib.h allows it to build and run.

#ifdef arc4random_stir
#undef arc4random_stir
#endif

I'm not a c genius, so not sure what implications this will have for me in the future. Primarily just wanted to try something that may help track down the real issue.

**Comment by edgarpettijohn** [comment]
The only issue I see thus far is:localhost$ sudo smtpd -d
info: OpenSMTPD 6.6.1-portable starting

^C[warn] event_del_: event has no event_base set.
Exiting

I can only seem to hit it during shutdown though.

**Comment by poolpOrg** [comment]
@edgarpettijohn you shouldn't try fixing things by editing system headers...

the issue is in the compat layer, very likely a check to add in configure.ac.

I'll be working on that next week.

**Comment by edgarpettijohn** [comment]
Playing around with it again. If I undefine HAVE_ARC4_RANDOM and use the openbsd-compat version it builds, installs, and runs fine.

**Comment by edgarpettijohn** [comment]
This seems to fix it for me.

diff --git a/openbsd-compat/openbsd-compat.h b/openbsd-compat/openbsd-compat.h
index 6c73e5b5..c7af0135 100644
--- a/openbsd-compat/openbsd-compat.h
+++ b/openbsd-compat/openbsd-compat.h
@@ -122,7 +122,7 @@ int getpeereid(int , uid_t *, gid_t *);
 unsigned int arc4random(void);
 #endif
 
-#if defined(HAVE_ARC4RANDOM_STIR)
+#if !defined(HAVE_ARC4RANDOM_STIR)
 void arc4random_stir(void);
 #elif defined(HAVE_ARC4RANDOM) || defined(LIBRESSL_VERSION_NUMBER)
 /* Recent system/libressl implementation; no need for explicit stir */

**Comment by poolpOrg** [comment]
I doubt this is the proper fix, it is likely going to break other builds.

Any chance you could lend me access to an account on a NetBSD box so I can debug this ?

My VPS providers don't provide NetBSD :-/

**Comment by ngortheone** [comment]
@poolpOrg 
I can get you one, AWS has 7.0 and 9.0RC1, no 8.1 unfortunately...

**Comment by poolpOrg** [comment]
He sent me an access by mail, currently troubleshooting 🙂

**Comment by poolpOrg** [comment]
The problem was in the `arc4random_stir()` detection but recent cleanups I wrote made it possible to plain remove a whole chunk of complex logic that led to the issue.

Using the latest portable commit, I can build and run correctly on your machine:
 
    netbsd$ nc localhost 25
    220 netbsd.pettijohn-web.com ESMTP OpenSMTPD
    ^C
    netbsd$
